Password Protected Posts Support

joehoyle opened the pull request:
In an attempt to better support password protected posts (see #2701) I've done a few things
now it's possible to better filter the password stuf in WordPress Core:
1. Password protected posts are now readable by every one.
2. All posts how have `content.protected` and `excerpt.protected` fields to indicate if those fields are protected by a password.
3. The `content.rendered` and `excerpt.rendered` are now empty strings if the post requires a password to be viewed.
4. Password protected posts in the rest api means "the post's content and excerpt are protected with a password".
5. If the user is viewing a post or listing posts in the `edit` context, they can see the protected fields.
6. The client can pass `?password=thepassword` to a single post endpoint to read it's protected fields.
7. An incorrect password with the above will result in a Forbidden response.
8. A client can now update and create password protected posts.
9. Post Types other than `post` no longer have support (or the property) for `password`.
10. Password protected post titles no longer have the "Protected: " prefix in REST API responses.

joehoyle committed:
Allow reading of password protected posts, and set the content
to an empty string if the user does not have access.

joehoyle committed:
Don't show "Protected: " on protected post titles in the rest api.

joehoyle committed:
Only test the existance for the password field for post_type===post

joehoyle committed:
Account for the fact the titles wont' have "Protected: " in them anymore

joehoyle committed:
Only test excerts and content for non-password protected posts.

This is pretty dificult to do anutomatically, as we don't have access to
the request vars so we don't know if the posts should have a valid
content or not.

joehoyle committed:
Update unit tests for new password protected posts handling.

kadamwhite commented:
Will test and review 9/23 AM by 12EDT

rmccue commented:
All 10 points here seem 👍 to me.

rmccue commented on lib/endpoints/class-wp-rest-posts-controller.php:
```diff
@@ -75,6 +78,11 @@ public function get_items_permissions_check( $request ) {
+		// Allow access to all password protected posts if the context is edit.
+		if ( 'edit' === $request['context'] ) {
+			add_filter( 'post_password_required', '__return_false' );
+		}
```
This could be combined into the above block by splitting it.

rmccue commented on lib/endpoints/class-wp-rest-posts-controller.php:
```diff
@@ -224,6 +232,20 @@ public function get_item_permissions_check( $request ) {
+			} else {
+				// Allow access to the post going forward.
+				add_filter( 'post_password_required', '__return_false' );
+			}
```
Unnecessary `else`

rmccue reviewed (changes requested):
The filters also need to be reset after use, otherwise the API isn't re-entrant.

kadamwhite commented on lib/endpoints/class-wp-rest-posts-controller.php:
```diff
@@ -265,10 +287,6 @@ public function create_item_permissions_check( $request ) {
 		$post_type = get_post_type_object( $this->post_type );
 
-		if ( ! empty( $request['password'] ) && ! current_user_can( $post_type->cap->publish_posts ) ) {
-			return new WP_Error( 'rest_cannot_publish', __( 'Sorry, you are not allowed to create password protected posts in this post type' ), array( 'status' => rest_authorization_required_code() ) );
```
Not 100% sure why this special case is needed, do password-protected posts have different capability requirements than regular posts? I'd not have expected to see a custom error message for the password-protected case here.

joehoyle replied:
Hmm, this is _no longer_ needed. It was here before to avoid support for password protected posts, which I now removed.

kadamwhite commented on lib/endpoints/class-wp-rest-posts-controller.php:
```diff
@@ -1137,34 +1111,30 @@ public function prepare_item_for_response( $post, $request ) {
 				/** This filter is documented in wp-includes/post-template.php */
-				'rendered' => apply_filters( 'the_content', $post->post_content ),
+				'rendered'  => post_password_required( $post ) ? '' : apply_filters( 'the_content', $post->post_content ),
+				'protected' => (bool) $post->post_password,
```
It's a little counterintuitive to me to have the `protected: true` enumerated in every `content` and `excerpt` object, _vs_ just having it declared on the root level. This feels like it's adding a lot of duplication in to every response object, especially given that `protected: false` is by far the most common case. Why not use a top-level `protected` key?

joehoyle replied:
The reason we put it under only these fields is to make it clear that it's only the `content` and the `excerpt` that is protected. The post it's self is not protected, hence it not being a property of the post object it's self. "protected posts" is a bit of a misnomer, "posts with protected content and excerpt". As for duplication, we'll need a `protected: false` on every post either way. However, it's true that (I think anyway) it's not possible to have an excerpt which is public and content which is private, however I'm not too concerned for the duplication. it's not really any more bytes over the wire with gzipping.

kadamwhite commented on lib/endpoints/class-wp-rest-posts-controller.php:
```diff
@@ -1137,34 +1111,30 @@ public function prepare_item_for_response( $post, $request ) {
-				'rendered' => $this->prepare_excerpt_response( $post->post_excerpt, $post ),
+				'raw'       => $post->post_excerpt,
+				'rendered'  => post_password_required( $post ) ? '' : $excerpt,
+				'protected' => (bool) $post->post_password,
```
I may have missed this, but I also want to understand the reasoning for `.protected` instead of `.password`, which would seem to better represent what's being asked for as well as better mirror the existing core post object

kadamwhite commented on lib/endpoints/class-wp-rest-posts-controller.php:
```diff
@@ -1244,6 +1214,20 @@ public function prepare_item_for_response( $post, $request ) {
+	 *
+	 * By default WordPress will show password protected posts with a title of
+	 * "Protected: %s", as the REST API communicates the protected status of a post
+	 * in a machine readable format, we remove the "Protected: " prefix.
```
Minor quibble: including "Protected: " has nothing to do with machine readability and everything to do with that string being purely presentational. I'd alter the explanation of this decision here.

joehoyle replied:
Sure can do - I was getting at there's no need to mark a post as "Protected: $x" because we communicate that it's protected in other ways.

kadamwhite replied:
:100: on the message, was just nitpicking the specifics :D

kadamwhite commented on lib/endpoints/class-wp-rest-posts-controller.php:
```diff
@@ -1619,6 +1610,12 @@ public function get_item_schema() {
+				'description' => __( 'A password to protect access to the content and excerpt.' ),
+				'type'        => 'string',
+				'context'     => array( 'edit' ),
+			);
```
I think I answered my question here... we can't have `.password` at the top level because that's where the string that sets that password must live. But I'm still 👎  on embedding the `.protected` key within the content and excerpt blocks; I'd prefer one of,
1. Have a `.password` property on the top level that is set as a string, and can be read as a string when password-authenticated, but otherwise displays as the boolean `true`; or
2. Have a `.protected` boolean property on the top level of the post object that represents whether the string `.password` field is present (whether a password is required)

joehoyle replied:
I'm 👎  on `password` being a string or a boolean depending on context. As I mentioned above, I don't think `protected` at the top level makes as much sense because it's not saying the whole post is protected, however I don't mind too much on that.

kadamwhite replied:
I'm curious whether anyone else agrees with me, then :) If not, I won't push.

Would we ever have content protected, but excerpt not?

kadamwhite commented on lib/endpoints/class-wp-rest-posts-controller.php:
```diff
@@ -1619,6 +1610,12 @@ public function get_item_schema() {
+				'description' => __( 'A password to protect access to the content and excerpt.' ),
+				'type'        => 'string',
+				'context'     => array( 'edit' ),
+			);
```
I think I answered my question here... we can't have `.password` at the top level because that's where the string that sets that password must live. But I'm still 👎  on embedding the `.protected` key within the content and excerpt blocks; I'd prefer one of,
1. Have a `.password` property on the top level that is set as a string, and can be read as a string when password-authenticated, but otherwise displays as the boolean `true`; or
2. Have a `.protected` boolean property on the top level of the post object that represents whether the string `.password` field is present (whether a password is required)

kadamwhite commented:
I confirmed the PR works as advertised, but had questions around the specifics of the implementation

rmccue commented:
Given that passwords affect more than one field, it makes sense to move `protected` out of the content/excerpt itself, I think. One reason we might not want to is if content and excerpt are independently protected, but I don't think that's possible?

rmccue committed:
Move post_password_required filtering to preparation

This allows resetting the filter after the content/excerpt have been
generated, ensuring the endpoints are re-entrant.

kadamwhite commented:
Missed @rmccue's comment, I had the same question about content/excerpt protection independence. Based on Ryan's comment I'd like to consolidate the two unless some specific use-case or possibility for separate protection states is demonstrated.

joehoyle committed:
Merge pull request #2735 from WP-API/move-password-filter

Move post_password_required filtering to preparation

rmccue commented:
I've changed my mind on this; I think `protected` inside the content/excerpt makes more sense, as it makes it much more obvious as to why the data is empty. I could see the potential for content and excerpt being protected separately as well, either in a plugin or in the future of core, and I think it's a little more future-proof.

joehoyle commented:
@rmccue your review still shows requesting changes. With your PR that is now merged into this, is there any more changes required from you?

joehoyle committed:
Fix tests for pages and attachments as they don't have the `password`
field anymore.

kadamwhite reviewed (approved):
I've been talked down on top-level `.protected` key, which is my one disagreement here; +1 from me on this

joehoyle committed:
We don't support password on pages and attachments no mo'

joehoyle committed:
Skip the test on < 4.7

joehoyle committed:
Merge branch 'develop' into password-protected-posts

joehoyle committed:
Fix yoda condition, I must.
